### PR TITLE
Check GitHub stars

### PR DESCRIPTION
**Pull request by klausenbusk**
Fix cdnjs/packages#299

### CODE REVIEW

**Comment by tc80** [comment]
Awesome idea! Maybe in another PR we can add a unit test for this, similar to the fake npm handler. (Will probably need to break `TestCheckerLint` into two different functions).

File: cmd/checker/main.go [hunk 1]
```diff
@@ -225,6 +226,28 @@ func makeGlobDebugLink(glob string, dir string) string {
 	return fmt.Sprintf("https://www.digitalocean.com/community/tools/glob?comments=true&glob=%s&matches=true%s&tests=", encodedGlob, allTests)
 }
 
+func checkPopularity(ctx context.Context, pckg *packages.Package) {
```
Comment (tc80) [inline cmd/checker/main.go:229]: do you think we need this function? seems like a lot of redundant checks for npm and git

**Comment by MattIPv4** [comment]
I thought we already had GitHub star checking for git auto-update packages, maybe not — 200 stars is what we’ve always used so this seems fine :)

**Comment by MattIPv4** [comment]
Does the longer already check GitHub stars from the auto update config if it is set to git? If not, should we also add that here?

**Comment by klausenbusk** [comment]
> Does the linter already check GitHub stars from the auto update config if it is set to git? If not, should we also add that here?

It always check GitHub stars if `pckg.Repository.URL` contains `github`.

File: cmd/checker/main.go (hunk 1 above)
Comment (klausenbusk) [reply cmd/checker/main.go:229]: We always want to check the GitHub stars (https://github.com/cdnjs/packages/issues/299). Do you have a smarter way in mind?

Comment (klausenbusk) [reply cmd/checker/main.go]: I think it should better now.

**Comment by thomasdavis** [comment]
What a cool update aha you guys are killing it

**Comment by MattIPv4** [comment]
> It always check GitHub stars if `pckg.Repository.URL` contains `github`.

Would there be a potential edge case where auto update has a repo set but the package doesn’t have a repo set, or where they have two different repos defined?

File: cmd/checker/main.go (hunk 1 above)
Comment (tc80) [reply cmd/checker/main.go]: aaaah, my bad!! didn't notice that issue - you're right

**Comment by klausenbusk** [comment]
> Would there be a potential edge case where auto update has a repo set but the package doesn’t have a repo set

The repository does not currently contains any packages like that:
```bash
$ find packages/ -name '*.json' | xargs -I{} -n1 -P 10 bash -c 'jq -e .repository.url {} > /dev/null || echo {}' | tee p
packages/t/tablefilter.json
packages/t/twilio.js.json
packages/f/flipCounter.json
packages/f/fatcow-icons.json
packages/e/extjs.json
packages/e/ext-core.json
packages/a/amstockchart.json
packages/d/dd_belatedpng.json
packages/j/jqplugin.json
packages/j/jquery.socialshareprivacy.json
packages/m/meyer-reset.json
packages/c/casualjs.json
packages/c/canvasjs.json
packages/c/chrome-frame.json
packages/l/log4javascript.json
packages/b/boexfi.json
packages/w/WebRupee.json
$ jq .autoupdate p
null
null
null
null
null
null
null
null
null
null
null
null
null
null
null
null
null
```

> or where they have two different repos defined?
```
$ jq 'if .autoupdate.source == "git" and (.repository.url | gsub(".*//|.git|/$"; "")) != (.autoupdate.target | gsub(".*//|.git|/$"; "")) then {"name": .name, "reporepo": .repository.url, "autorepo": .autoupdate.target} else {} end' $(find packages/ -name '*.json') | grep -v '{}'
{
  "name": "tuesday",
  "reporepo": "https://github.com/ShakrMedia/tuesday/",
  "autorepo": "https://github.com/shakrmedia/tuesday.git"
}
{
  "name": "timezone-js",
  "reporepo": "git://github.com/mde/timezone-js.git",
  "autorepo": "git://github.com/imakewebthings/waypoints.git"
}
{
  "name": "zepto.fullpage",
  "reporepo": "git@github.com:yanhaijing/zepto.fullpage.git",
  "autorepo": "git://github.com/yanhaijing/zepto.fullpage.git"
}
{
  "name": "zoom.js",
  "reporepo": "git@github.com:fat/zoom.js.git",
  "autorepo": "git://github.com/fat/zoom.js.git"
}
{
  "name": "zabuto_calendar",
  "reporepo": "git@github.com:zabuto/calendar.git",
  "autorepo": "git://github.com/zabuto/calendar.git"
}
{
  "name": "primeui",
  "reporepo": "https://github.com/primefaces/primeui",
  "autorepo": "git://github.com/primefaces/primeui-distribution.git"
}
{
  "name": "PKI.js",
  "reporepo": "git://github.com/PeculiarVentures/PKI.js.git",
  "autorepo": "git://github.com/GlobalSign/PKI.js.git"
}
{
  "name": "porthole",
  "reporepo": "https://ternarylabs.github.io/porthole/",
  "autorepo": "https://github.com/ternarylabs/porthole.git"
}
{
  "name": "polyglot",
  "reporepo": "git@github.com:tavyy/Polyglot-Language-Switcher.git",
  "autorepo": "git://github.com/tavyy/Polyglot-Language-Switcher.git"
}
{
  "name": "ember-simple-auth",
  "reporepo": "https://github.com/simplabs/ember-simple-auth",
  "autorepo": "git://github.com/simplabs/ember-simple-auth-component.git"
}
{
  "name": "ember-data.js",
  "reporepo": "https://github.com/emberjs/data",
  "autorepo": "git://github.com/components/ember-data.git"
}
{
  "name": "ember.js",
  "reporepo": "https://github.com/emberjs/ember.js",
  "autorepo": "git://github.com/components/ember.git"
}
{
  "name": "elemental",
  "reporepo": "https://github.com/elementalui/elemental.git",
  "autorepo": "git://github.com/elementalui/elemental-dist.git"
}
{
  "name": "equalize.js",
  "reporepo": "https://github.com/tsvensen/equalize.js/blob/master/js/equalize.js",
  "autorepo": "git://github.com/tsvensen/equalize.js.git"
}
{
  "name": "nivo-lightbox",
  "reporepo": "git@github.com:gilbitron/Nivo-Lightbox.git",
  "autorepo": "git://github.com/gilbitron/Nivo-Lightbox.git"
}
{
  "name": "angular-timer",
  "reporepo": "git@github.com:siddii/angular-timer.git",
  "autorepo": "git://github.com/siddii/angular-timer.git"
}
{
  "name": "angularFire",
  "reporepo": "git://github.com/firebase/angularFire.git",
  "autorepo": "git://github.com/firebase/angularfire.git"
}
{
  "name": "angular-foundation",
  "reporepo": "git://github.com/pineconellc/angular-foundation.git",
  "autorepo": "git://github.com/pineconellc/angular-foundation-bower.git"
}
{
  "name": "autobahn",
  "reporepo": "git://github.com/crossbario/autobahn-js.git",
  "autorepo": "git://github.com/crossbario/autobahn-js-browser.git"
}
{
  "name": "alchemyjs",
  "reporepo": "https://github.com/graphalchemist/alchemy.git",
  "autorepo": "git://github.com/GraphAlchemist/Alchemy.git"
}
{
  "name": "alertifyjs-alertify.js",
  "reporepo": "git@github.com:alertifyjs/alertify.js.git",
  "autorepo": "git://github.com/alertifyjs/alertify.js.git"
}
{
  "name": "gifffer",
  "reporepo": "git@github.com:krasimir/gifffer.git",
  "autorepo": "git://github.com/krasimir/gifffer.git"
}
{
  "name": "gator",
  "reporepo": "git@github.com:ccampbell/gator.git",
  "autorepo": "git://github.com/ccampbell/gator.git"
}
{
  "name": "kurento-client",
  "reporepo": "https://github.com/Kurento/kurento-client-js.git",
  "autorepo": "git://github.com/Kurento/kurento-client-bower.git"
}
{
  "name": "darkroomjs",
  "reporepo": "https://github.com/Mattketmo/darkroomjs.git",
  "autorepo": "https://github.com/MattKetmo/darkroomjs.git"
}
{
  "name": "react-uwp",
  "reporepo": "git://github.com/myxvisual/react-uwp.git",
  "autorepo": "git://github.com/myxvisual/react-uwp-bower.git"
}
{
  "name": "rantjs",
  "reporepo": "https://github.com/svenanders/rantjs.git",
  "autorepo": "git://github.com/svenanders/Rantjs.git"
}
{
  "name": "jquery.dirtyforms.helpers.tinymce",
  "reporepo": "https://github.com/snikch/jquery.dirtyforms.git",
  "autorepo": "git://github.com/NightOwl888/jquery.dirtyforms.helpers.tinymce.dist.git"
}
{
  "name": "jquery.dirtyforms.helpers.ckeditor",
  "reporepo": "https://github.com/snikch/jquery.dirtyforms.git",
  "autorepo": "git://github.com/NightOwl888/jquery.dirtyforms.helpers.ckeditor.dist.git"
}
{
  "name": "jquery.dirtyforms.helpers.alwaysdirty",
  "reporepo": "https://github.com/snikch/jquery.dirtyforms.git",
  "autorepo": "git://github.com/NightOwl888/jquery.dirtyforms.helpers.alwaysdirty.dist.git"
}
{
  "name": "jquery.dirtyforms.dialogs.pnotify",
  "reporepo": "https://github.com/snikch/jquery.dirtyforms.git",
  "autorepo": "git://github.com/NightOwl888/jquery.dirtyforms.dialogs.pnotify.dist.git"
}
{
  "name": "JQuery-Snowfall",
  "reporepo": "https://loktar00@github.com/loktar00/JQuery-Snowfall.git",
  "autorepo": "git://github.com/loktar00/JQuery-Snowfall.git"
}
{
  "name": "jquery-lang-js",
  "reporepo": "git://github.com/irrelon/jquery-lang-js.git",
  "autorepo": "git://github.com/Irrelon/jquery-lang-js.git"
}
{
  "name": "jquery.dirtyforms.dialogs.jquery-ui",
  "reporepo": "https://github.com/snikch/jquery.dirtyforms.git",
  "autorepo": "git://github.com/NightOwl888/jquery.dirtyforms.dialogs.jquery-ui.dist.git"
}
{
  "name": "jquery.dirtyforms",
  "reporepo": "https://github.com/snikch/jquery.dirtyforms.git",
  "autorepo": "git://github.com/NightOwl888/jquery.dirtyforms.dist.git"
}
{
  "name": "jquery.dirtyforms.dialogs.facebox",
  "reporepo": "https://github.com/snikch/jquery.dirtyforms.git",
  "autorepo": "git://github.com/NightOwl888/jquery.dirtyforms.dialogs.facebox.dist.git"
}
{
  "name": "jquery.dirtyforms.dialogs.blockui",
  "reporepo": "https://github.com/snikch/jquery.dirtyforms.git",
  "autorepo": "git://github.com/NightOwl888/jquery.dirtyforms.dialogs.blockui.dist.git"
}
{
  "name": "jquery.dirtyforms.dialogs.bootstrap",
  "reporepo": "https://github.com/snikch/jquery.dirtyforms.git",
  "autorepo": "git://github.com/NightOwl888/jquery.dirtyforms.dialogs.bootstrap.dist.git"
}
{
  "name": "jInvertScroll",
  "reporepo": "git@github.com:pixxelfactory/jInvertScroll.git",
  "autorepo": "git://github.com/pixxelfactory/jInvertScroll.git"
}
{
  "name": "jquery.cookieBar",
  "reporepo": "git@github.com:carlwoodhouse/jquery.cookieBar.git",
  "autorepo": "git://github.com/carlwoodhouse/jquery.cookieBar.git"
}
{
  "name": "metro",
  "reporepo": "git://github.com:olton/Metro-UI-CSS.git",
  "autorepo": "git://github.com/olton/Metro-UI-CSS.git"
}
{
  "name": "highmaps",
  "reporepo": "https://github.com/highcharts/highcharts.git",
  "autorepo": "git://github.com/highslide-software/highmaps-release.git"
}
{
  "name": "highcharts",
  "reporepo": "https://github.com/highcharts/highcharts.git",
  "autorepo": "git://github.com/highcharts/highcharts-dist.git"
}
{
  "name": "highlight.js",
  "reporepo": "https://github.com/isagalaev/highlight.js",
  "autorepo": "git://github.com/highlightjs/cdn-release.git"
}
{
  "name": "highchartTable",
  "reporepo": "git@github.com:highchartTable/jquery-highchartTable-plugin.git",
  "autorepo": "git://github.com/highchartTable/jquery-highchartTable-plugin.git"
}
{
  "name": "highstock",
  "reporepo": "https://github.com/highcharts/highcharts",
  "autorepo": "git://github.com/highslide-software/highstock-release.git"
}
{
  "name": "openseadragon",
  "reporepo": "https://github.com/openseadragon/openseadragon.git",
  "autorepo": "git://github.com/openseadragon/site-build.git"
}
{
  "name": "vertx",
  "reporepo": "https://github.com/vert-x3/vertx-web",
  "autorepo": "git://github.com/vert-x3/vertx-bus-bower.git"
}
{
  "name": "social-share-kit",
  "reporepo": "git@github.com:darklow/social-share-kit.git",
  "autorepo": "git://github.com/darklow/social-share-kit.git"
}
{
  "name": "s3colors",
  "reporepo": "git://github.com/shaz3e/S3-Colors.git",
  "autorepo": "git://github.com/Shaz3e/S3-Colors.git"
}
{
  "name": "slideout",
  "reporepo": "git://github.com/mango/slideout.git",
  "autorepo": "git://github.com/Mango/slideout.git"
}
{
  "name": "spark-md5",
  "reporepo": "git@github.com:satazor/SparkMD5.git",
  "autorepo": "git://github.com/satazor/SparkMD5.git"
}
{
  "name": "smoothState.js",
  "reporepo": "git@github.com:miguel-perez/smoothState.js.git",
  "autorepo": "git://github.com/miguel-perez/smoothState.js.git"
}
{
  "name": "css-element-queries",
  "reporepo": "git@github.com:marcj/css-element-queries.git",
  "autorepo": "git://github.com/marcj/css-element-queries.git"
}
{
  "name": "chosen",
  "reporepo": "https://github.com/harvesthq/chosen.git",
  "autorepo": "git://github.com/harvesthq/chosen-package.git"
}
{
  "name": "l20n",
  "reporepo": "https://github.com/l20n/l20n.js",
  "autorepo": "git://github.com/l20n/builds.git"
}
{
  "name": "logerr",
  "reporepo": "git@github.com:i-break-codes/logerr.git",
  "autorepo": "git://github.com/i-break-codes/logerr.git"
}
{
  "name": "Blotter",
  "reporepo": "https://github.com/bradley/blotter",
  "autorepo": "git://github.com/bradley/Blotter.git"
}
{
  "name": "bxslider",
  "reporepo": "https://github.com/stevenwanderski/bxslider-4/bxslider-4.git",
  "autorepo": "git://github.com/stevenwanderski/bxslider-4.git"
}
{
  "name": "Bridge.NET",
  "reporepo": "git://github.com/bridgedotnet/Bridge.git",
  "autorepo": "git://github.com/bridgedotnet/Speed.git"
}
{
  "name": "bootstrap-datetimepicker",
  "reporepo": "git://github.com/eonasdan/bootstrap-datetimepicker.git",
  "autorepo": "git://github.com/Eonasdan/bootstrap-datetimepicker.git"
}
{
  "name": "winjs",
  "reporepo": "https://github.com/winjs/winjs",
  "autorepo": "git://github.com/winjs/winjs-bower.git"
}
```

@MattIPv4 neverthelss adding a check should be simple enough:
```go
if pckg.Repository.URL == nil {
    err(ctx, shouldNotBeEmpty(".repository.url"))
}
if pckg.Autoupdate.Source == "git" && pckg.Autoupdate.Target != pckg.Repository.URL {
    err(ctx, ".autoupdate.target and repository.url must not differ")
}
```

**Comment by klausenbusk** [comment]
> @MattIPv4 neverthelss adding a check should be simple enough:

Fixed

File: git/github.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	"github.com/cdnjs/tools/util"
+)
+
+// Stars holds the number of stars for a GitHub repository.
+type Stars struct {
+	Stars uint `json:"stargazers_count"`
+}
+
+// Gets the protocol, either http or https.
+func getProtocol() string {
```
Comment (xtuc) [inline git/github.go:18]: could we maybe shared this npm code?

Comment (klausenbusk) [reply git/github.go]: Like this? or is there a better way to do it?

**Comment by klausenbusk** [comment]
> can you please add a test that triggers the GitHub min? Like we did for npm

Done

File: test/checker/lint_test.go [hunk 3]
```diff
@@ -21,11 +21,13 @@ const (
...
 	nonexistentPkg = "nonexistent"
 	normalPkg      = "normal"
+	unpopularRepo  = "user/unpopularRepo"
 )
 
-// fakes the npm api for testing purposes
-func fakeNpmHandlerLint(w http.ResponseWriter, r *http.Request) {
+// fakes the npm api and GitHub api for testing purposes
+func fakeNpmGitHubHandlerLint(w http.ResponseWriter, r *http.Request) {
 	switch r.URL.Path {
```
Comment (xtuc) [inline test/checker/lint_test.go:29]: to make it clearer between github and npm we could concatenate the Host header and the path

Comment (klausenbusk) [reply test/checker/lint_test.go]: Good idea, done!

**Pull request merged**